stl/inc/variant

```diff
@@ -1601,6 +1626,51 @@ template <class _Ty>
 
+#if _HAS_CXX20
+template <size_t _Size, class _ListOfIndexVectors, class _Ret, class _Callable, class... _Variants>
+constexpr _Ret _Visit_impl(_Callable&& _Obj, _Variants&&... _Args) {
+    constexpr int _Strategy =
+        _Size == 1 ? 0 : _Size <= 4 ? 1 : _Size <= 16 ? 2 : _Size <= 64 ? 3 : _Size <= 256 ? 4 : -1;
+    return _Visit_strategy<_Strategy>::template _Visit<_Ret, _ListOfIndexVectors>(
+        _Variant_visit_index1(0, static_cast<_As_variant<_Variants>&>(_Args)...), static_cast<_Callable&&>(_Obj),
+        static_cast<_As_variant<_Variants>&&>(_Args)...);
+}
+
+template <class _Callable, class... _Variants, class = void_t<_As_variant<_Variants>...>>
+constexpr _Variant_visit_result_t<_Callable, _As_variant<_Variants>...> visit(_Callable&& _Obj, _Variants&&... _Args) {
+    // Invoke _Obj with the contained values of _Args...
+    constexpr auto _Size = _Variant_total_states<_Remove_cvref_t<_As_variant<_Variants>>...>;
+    using _ListOfIndexLists =
+        _Meta_list<_Meta_as_list<make_index_sequence<1 + variant_size_v<_Remove_cvref_t<_As_variant<_Variants>>>>>...>;
+    using _ListOfIndexVectors =
+        _Meta_transform<_Meta_quote<_Meta_as_integer_sequence>, _Meta_cartesian_product<_ListOfIndexLists>>;
+    static_assert(_Variant_all_visit_results_same<_Callable, _ListOfIndexVectors, _As_variant<_Variants>...>::value,
+        "visit() requires the result of all potential invocations to have the same type and value category "
+        "(N4828 [variant.visit]/2).");
+
+    return _Visit_impl<_Size, _ListOfIndexVectors, _Variant_visit_result_t<_Callable, _As_variant<_Variants>...>>(
+        _STD forward<_Callable>(_Obj), _STD forward<_Variants>(_Args)...);
+}
+
+template <class _Ret, class _Callable, class... _Variants, class = void_t<_As_variant<_Variants>...>>
+constexpr _Ret visit(_Callable&& _Obj, _Variants&&... _Args) {
+    // Invoke _Obj with the contained values of _Args...
+    constexpr auto _Size = _Variant_total_states<_Remove_cvref_t<_As_variant<_Variants>>...>;
+    using _ListOfIndexLists =
+        _Meta_list<_Meta_as_list<make_index_sequence<1 + variant_size_v<_Remove_cvref_t<_As_variant<_Variants>>>>>...>;
+    using _ListOfIndexVectors =
+        _Meta_transform<_Meta_quote<_Meta_as_integer_sequence>, _Meta_cartesian_product<_ListOfIndexLists>>;
+    if constexpr (!is_void_v<_Ret>) {
+        static_assert(_Variant_all_visit_results_implicitly_convertible<_Ret, _Callable, _ListOfIndexVectors,
+                          _As_variant<_Variants>...>::value,
+            "visit<R>() requires the result of all potential invocations to be implicitly convertible to R "
+            "(N4828 [variant.visit]/2).");
+    }
+
+    return _Visit_impl<_Size, _ListOfIndexVectors, _Ret>(
+        _STD forward<_Callable>(_Obj), _STD forward<_Variants>(_Args)...);
+}
+#else // _HAS_CXX20
 template <class _Callable, class... _Variants, class = void_t<_As_variant<_Variants>...>>

```

Comment:
IIUC this version of vanilla visit is identical to the version in the `_HAS_CXX20` block above, except that `_Visit_impl` is effectively inlined into this one. If that's the case, we could reduce duplication by deleting this version and moving the other version and `_Visit_impl` out of the `_HAS_CXX20` block so they're used in both language modes.